### PR TITLE
[BTRFS] update driver to latest upstream version (1.8.0)

### PR DESCRIPTION
**Pull request by Vince1171**
JIRA issue: [CORE-17076](https://jira.reactos.org/browse/CORE-17076)

## Purpose

+ Keep us as close as possible from upstream. (https://github.com/maharmstone/btrfs)
+ (Allow me to get familiar with ReactOS code, it's my first contribution...)

## Proposed changes
+ update the driver ^^
+ cleanup ReactOS speciffic code
+ import BitLocker NTStatus
### Changelog from upstream
v1.8 (2022-03-12):
* Added minimal support for fs-verity
* ~Added test suite~ **not in ReactOS**
* Fixed incorrect disk usage statistics
* Fixed potential crashes when renaming stream to file or file to stream
* Fixed potential crashes when querying hard links on file
* Fixed potential hang when opening oplocked file
* Fixed minor issues also uncovered by test suite

v1.7.9 (2021-10-02):
* Fixed deadlock when mounting on Windows 11
* Added support for BitLocker-encrypted volumes
* Improved filename checks when renaming or creating hard links
* Miscellaneous bug fixes

v1.7.8.1 (2021-06-13):
* Fixed bug preventing new directories from appearing in listings
* Fixed Release version of driver still not working on XP

v1.7.8 (2021-06-09):
* ~Upgraded zstd to version 1.5.0~ **not in ReactOS as it introduce deprecated code**
* Fixed regression stopping driver from working under XP
* Fixed compilation on clang
* Fixed corruption issue when Linux mount option `inode_cache` had been used
* Fixed recursion issue involving virtual directory \\$Root

v1.7.7 (2021-04-12):
* Fixed deadlock on high load
* Fixed free space issue when installing Genshin Impact
* Fixed issue when copying files with wildcards in command prompt
* Increased speed of directory lookups

v1.7.6 (2021-01-14):
* Fixed race condition when booting with Quibble
* No longer need to restart Windows after initial installation
* Forced maximum file name to 255 UTF-8 characters, to match Linux driver
* Fixed issue where directories could be created with trailing backslash
* Fixed potential deadlock when Windows calls NtCreateSection during flush
* Miscellaneous bug fixes

v1.7.5 (2020-10-31):
* Fixed text display issue in shell extension
* Added support for mingw 8
* Fixed LXSS permissions not working in new versions of Windows
* Fixed issue where truncating an inline file wouldn't change its size
* Fixed crash with Quibble where driver would try to use AVX2 before Windows had enabled it

v1.7.4 (2020-08-23):
* Fixed issue when running compressed EXEs
* Changed build system to cmake
* Upgraded zstd to version 1.4.5
* Added support for FSCTL_GET_RETRIEVAL_POINTERS
* Miscellaneous bug fixes

v1.7.3 (2020-05-24):
* Fixed crash when sending file change notifications
* Improved symlink handling with LXSS
* Added support for undocumented flag SL_IGNORE_READONLY_ATTRIBUTE
* Fixed corruption caused by edge case, where address allocated and freed in same flush
* Improved handling of free space tree
* Improved handling of very full volumes
* Fixed spurious warnings raised by GCC 10 static analyser
* Replaced multiplications and divisions with bit shift operations where appropriate
* Fixed combobox stylings in shell extension

### CODE REVIEW

File: sdk/lib/fslib/btrfslib/btrfslib.c [hunk 1]
```diff
@@ -1263,13 +1261,6 @@ static bool is_mounted_multi_device(HANDLE h, uint32_t sector_size) {
         bfs = malloc(bfssize);
 
         Status = NtDeviceIoControlFile(h2, NULL, NULL, NULL, &iosb, IOCTL_BTRFS_QUERY_FILESYSTEMS, NULL, 0, bfs, bfssize);
-        if (!NT_SUCCESS(Status) && Status != STATUS_BUFFER_OVERFLOW) {
```
Comment (HBelusca) [inline sdk/lib/fslib/btrfslib/btrfslib.c:1266]: Is it certain this is not necessary anymore?

File: drivers/filesystems/btrfs/volume.c [hunk 2]
```diff
@@ -877,8 +812,9 @@ NTSTATUS vol_device_control(IN PDEVICE_OBJECT DeviceObject, IN PIRP Irp) {
             return vol_get_disk_extents(vde, Irp);
 
         default: { // pass ioctl through if only one child device
-            ULONG code = IrpSp->Parameters.DeviceIoControl.IoControlCode;
             NTSTATUS Status = vol_ioctl_passthrough(vde, Irp);
+#ifdef _DEBUG
+            ULONG code = IrpSp->Parameters.DeviceIoControl.IoControlCode;
 
 #ifdef __REACTOS__
```
Comment (HBelusca) [inline drivers/filesystems/btrfs/volume.c:819]: It looks that this ifdef block could be removed now...

File: drivers/filesystems/btrfs/treefuncs.c [hunk 3]
```diff
@@ -18,6 +18,7 @@
 #include "btrfs_drv.h"
 #include "crc32c.h"
 
+__attribute__((nonnull(1,3,4,5)))
```
Comment (HBelusca) [inline drivers/filesystems/btrfs/treefuncs.c:21]: lol.

File: drivers/filesystems/btrfs/btrfs.h [hunk 4]
```diff
@@ -9,12 +9,14 @@
 #pragma once
 
 #include <stdint.h>
+#include <assert.h>
```
Comment (HBelusca) [inline drivers/filesystems/btrfs/btrfs.h:12]: Is that (--> using that header) allowed in kernel mode?

File: dll/win32/ubtrfs/CMakeLists.txt [hunk 5]
```diff
@@ -3,10 +3,9 @@ spec2def(ubtrfs.dll ubtrfs.spec)
 
 list(APPEND SOURCE
     ubtrfs.c
-    ubtrfs.rc
     ${CMAKE_CURRENT_BINARY_DIR}/ubtrfs.def)
 
-add_library(ubtrfs MODULE ${SOURCE})
```
Comment (HBelusca) [inline dll/win32/ubtrfs/CMakeLists.txt:9]: Is this change necessary?

File: dll/win32/ubtrfs/ubtrfs.rc [hunk 6]
```diff
@@ -25,18 +25,18 @@ LANGUAGE LANG_ENGLISH, SUBLANG_ENGLISH_UK
 // TEXTINCLUDE
```
Comment (HBelusca) [inline dll/win32/ubtrfs/ubtrfs.rc:25]: What is happening with the file rights?
image

File: dll/shellext/shellbtrfs/shellbtrfs.rc [hunk 7]
```diff
@@ -9,6 +9,7 @@
 //
```
Comment (HBelusca) [inline dll/shellext/shellbtrfs/shellbtrfs.rc:9]: Similarly here there's a change of file rights or whatever...

File: dll/shellext/shellbtrfs/recv.cpp [hunk 8]
```diff
@@ -1736,6 +1645,11 @@ void CALLBACK RecvSubvolW(HWND hwnd, HINSTANCE hinst, LPWSTR lpszCmdLine, int nC
 
             tp->PrivilegeCount = 3;
 
+#ifndef __REACTOS__
```
Comment (HBelusca) [inline dll/shellext/shellbtrfs/recv.cpp:1648]: The WinBtrfs author doesn't compile his code with MSVC too? (aka. would he have then noticed whether these pragmas are just ignored there, or would error out as being unrecognized?)

File: drivers/filesystems/btrfs/btrfs.c [hunk 9]
```diff
@@ -627,6 +668,9 @@ static bool lie_about_fs_type() {
     INIT_UNICODE_STRING(fsutil, L"FSUTIL.EXE");
     INIT_UNICODE_STRING(storsvc, L"STORSVC.DLL");
 
+    /* Not doing a Volkswagen, honest! Some IFS tests won't run if not recognized FS. */
```
Comment (HBelusca) [inline drivers/filesystems/btrfs/btrfs.c:671]: lmao 😂 
Btw here I'm attaching URLs about that "IFSTest" thingie for reference purposes:
https://docs.microsoft.com/en-us/previous-versions/windows/hardware/wlk/ff563405(v=vs.85)
https://docs.microsoft.com/en-us/windows-hardware/test/hlk/testref/14b230f3-7eee-437e-ab2f-375b200de6f3
https://docs.microsoft.com/en-us/windows-hardware/test/hlk/testref/e84242de-56a9-4e6b-8f8c-ad3f7252cec7

File: drivers/filesystems/btrfs/crc32c-gas.S [hunk 10]
```diff
@@ -17,8 +17,105 @@
 
 #include <asm.inc>
 
-EXTERN _crctable:DWORD
```
Comment (HBelusca) [inline drivers/filesystems/btrfs/crc32c-gas.S:20]: Question for these ASM files: Looking at upstream winbtrfs there are separate MASM and GAS files. Here you've kept our synthetic ASM file that can be compiled at the same time on both MASM and GAS...
So what's the best plan here? Keep that file, or switch back to the separate files?

**Comment by HBelusca** [comment]
> ~~Upgraded zstd to version 1.5.0~~ **not in ReactOS as it introduce deprecated code**

Any details about what this deprecated code is?

**Comment by HBelusca** [comment]
Have you tested whether ReactOS still installs, boots, and run fine with this updated version?

**Review by JoachimHenze — CHANGES_REQUESTED** [review]
You chose a complex topic for first contribution.
Unfortunately not are compilers are happy with your proposed changes yet, e.g.:
[build-msvc (14.2, i386, Debug)](https://github.com/reactos/reactos/runs/5762004702?check_suite_focus=true#logs) fails with:

D:\a\reactos\reactos\build\drivers\filesystems\btrfs\asm\xor-gas_btrfs_asm.asm(719) : error A2070:invalid instruction operands
D:\a\reactos\reactos\build\drivers\filesystems\btrfs\asm\xor-gas_btrfs_asm.asm(720) : error A2070:invalid instruction operands
D:\a\reactos\reactos\build\drivers\filesystems\btrfs\asm\xor-gas_btrfs_asm.asm(722) : error A2070:invalid instruction operands

Thank you still, for your work so far.

File: dll/shellext/shellbtrfs/main.cpp [hunk 11]
```diff
@@ -499,7 +497,7 @@ static void unreg_prop_sheet_handler(const wstring& filetype, const wstring& nam
 #endif
 }
 
-extern "C" STDAPI DllRegisterServer(void) {
+STDAPI DllRegisterServer(void) {
```
Comment (learn-more) [inline dll/shellext/shellbtrfs/main.cpp:500]: I am pretty sure that this won't work like this anymore, did you test it?

File: drivers/filesystems/btrfs/volume.c (hunk 2 above)
Comment (Vince1171) [reply drivers/filesystems/btrfs/volume.c:819]: indeed

File: drivers/filesystems/btrfs/btrfs.h (hunk 4 above)
Comment (Vince1171) [reply drivers/filesystems/btrfs/btrfs.h:12]: it seems I forgot to surround it with ifdef like the static_assert at line 306

File: dll/win32/ubtrfs/ubtrfs.rc (hunk 6 above)
Comment (Vince1171) [reply dll/win32/ubtrfs/ubtrfs.rc:25]: as said here above, it's a mistake, I'll change it back asap

File: dll/shellext/shellbtrfs/recv.cpp (hunk 8 above)
Comment (Vince1171) [reply dll/shellext/shellbtrfs/recv.cpp:1648]: I'll setup MSVC to test.
but WinMG on Linux is not happy about it.

File: drivers/filesystems/btrfs/crc32c-gas.S (hunk 10 above)
Comment (Vince1171) [reply drivers/filesystems/btrfs/crc32c-gas.S:20]: here we have 2 possibilities:
 * match upstream and have separate files for MASM and GAS.
 * continue to use our "synthetic" one.
I currently used the second one, but I have to admit the first one is
1) easier to maintain
2) closer to upstream

I'll change it to 2 separate files

**Comment by Vince1171** [comment]
> Have you tested whether ReactOS still installs, boots, and run fine with this updated version?

ReactOS boots, installs and run fine at each commit of this PR.
But I have to admit I only compiled it with winMG as I run Linux...
MSVC seems to be broken so I'll setup a Windows env and fix all MSVC issues asap :)

File: dll/shellext/shellbtrfs/main.cpp (hunk 11 above)
Comment (Vince1171) [reply dll/shellext/shellbtrfs/main.cpp:500]: by definition STDAPI implies extern "C". Or did I miss something? 
and yes it's tested and working

Comment (learn-more) [reply dll/shellext/shellbtrfs/main.cpp:500]: Ah right, it is indeed part of that already.
Without the `extern "C"` thing gcc usually fails to act normal.

**Comment by HBelusca** [comment]
There are some residual build failures, please double-check.

File: dll/win32/ubtrfs/ubtrfs.rc (hunk 6 above)
Comment (Vince1171) [reply dll/win32/ubtrfs/ubtrfs.rc:25]: so after checking it seems all files permissions are set to 755 except *.rc, *CMakeLists.txt, and *.S files, any idea why?

Comment (Vince1171) [reply dll/win32/ubtrfs/ubtrfs.rc:25]: and it seems to be only for btrfs driver, I checked other drivers files, and they are all set to 755
it may be an error from the previous import of the btrfs driver?

File: sdk/lib/fslib/btrfslib/btrfslib.c (hunk 1 above)
Comment (Vince1171) [reply sdk/lib/fslib/btrfslib/btrfslib.c:1266]: according to upstream, it has been removed to fix a memory leak  
https://github.com/maharmstone/btrfs/commit/6ceb31763a285fc45cc6814b9fea35be5c84c306  
but the leak seems to have been managed in ReactOS with the `if (bfs) free(bfs);`  
but upstream hasn't been very explicit in his commit message... so I don't have info than that...

File: drivers/filesystems/btrfs/btrfs.h (hunk 4 above)
Comment (Vince1171) [reply drivers/filesystems/btrfs/btrfs.h:13]: done in e720186ed8beb765c75222249dfe7e6d78bb3ab8

File: drivers/filesystems/btrfs/volume.c (hunk 2 above)
Comment (Vince1171) [reply drivers/filesystems/btrfs/volume.c]: changed in e720186ed8beb765c75222249dfe7e6d78bb3ab8

File: dll/win32/ubtrfs/CMakeLists.txt (hunk 5 above)
Comment (Vince1171) [reply dll/win32/ubtrfs/CMakeLists.txt]: changed in e720186ed8beb765c75222249dfe7e6d78bb3ab8

File: dll/shellext/shellbtrfs/recv.cpp (hunk 8 above)
Comment (Vince1171) [reply dll/shellext/shellbtrfs/recv.cpp]: gcc was the one not happy due to the -Wall adding -Wunknown-pragmas
changed to `ifdef __clang__`.

File: drivers/filesystems/btrfs/crc32c-gas.S (hunk 10 above)
Comment (Vince1171) [reply drivers/filesystems/btrfs/crc32c-gas.S]: I changed the name of the *.S to remove the "gas" from the name to avoid confusion.
compilation for MSVC has been fixed

**Comment by julenuri** [comment]
It works for me under GCC. I didn't test under x64 or other compilers version. Some conflicts running some commands under CMD, but that happened in the past.

**Comment by Vince1171** [comment]
> > ~Upgraded zstd to version 1.5.0~ **not in ReactOS as it introduce deprecated code**
> 
> Any details about what this deprecated code is?

I didn't describe it correctly   
Starting from 1.5, the zstd methods used by Winbtrfs have been flagged as deprecated...  
Upstream simply compile without the depreciation warning.  I choosed to stay with 1.4...
Should I follow the same path as upstream and silence the deprecation warning?  
  
Otherwise, this PR is ready to be merged. All builds have been fixed.

File: drivers/filesystems/btrfs/xor-gas.S [hunk 12]
```diff
@@ -116,11 +116,11 @@ do_xor_avx2:
 do_xor_avx2_loop:
     cmp r8d, 32
     jl stragglers4
-
-    vmovdqa ymm0, [rcx]
-    vmovdqa ymm1, [rdx]
+
```
Comment (HBelusca) [inline drivers/filesystems/btrfs/xor-gas.S]:
```suggestion

```

**Review by HBelusca — APPROVED** [review]

File: dll/shellext/shellbtrfs/contextmenu.cpp [hunk 13]
```diff
@@ -1719,8 +1715,4 @@ void CALLBACK ReflinkCopyW(HWND hwnd, HINSTANCE hinst, LPWSTR lpszCmdLine, int n
             }
         }
     }
-}
-
-#ifdef __REACTOS__
-} /* extern "C" */
-#endif
+}
```
Comment (HBelusca) [inline dll/shellext/shellbtrfs/contextmenu.cpp:1718]: sus

File: dll/shellext/shellbtrfs/shellbtrfs.rc [hunk 14]
```diff
@@ -9,6 +9,7 @@
...
 #include "winres.h"
 
+
```
Comment (HBelusca) [inline dll/shellext/shellbtrfs/shellbtrfs.rc:12]:
```suggestion
```

**Review by HBelusca — APPROVED** [review]

**Comment by HBelusca** [comment]
Regarding the file mode changes, please see https://tecadmin.net/git-ignore-file-permission/ and disable that thing in your git settings.

**Comment by HBelusca** [comment]
I've reorganized the "fix-commits" (that followed the original update-commits) and squashed their contents with their respective update-commits, so that each resulting individual update-commit successfully builds on all builders.
I will investigate the usage of SEH in this code base tomorrow, before re-approving the PR.

**Review by HBelusca — APPROVED** [review]

**Comment by HBelusca** [comment]
Thanks for your PR!
